app/src/main/java/org/septa/android/app/favorites/FavoritesFragment.java

```diff
@@ -117,42 +126,103 @@ private String evaluateAndRemoveFavorites(Map<String, Favorite> favoritesMap) {
         return msg;
     }
 
-    private View onCreateViewFavorites(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+    private View onCreateViewFavorites(LayoutInflater inflater, final ViewGroup container, Bundle savedInstanceState) {
         setHasOptionsMenu(true);
         fragmentView = inflater.inflate(R.layout.favorites_view, container, false);
         LinearLayout favoritesListView = (LinearLayout) fragmentView.findViewById(R.id.favorites_list);
 
-        for (Map.Entry<String, Favorite> entry : favoritesMap.entrySet()) {
-            final Favorite favorite = entry.getValue();
+        // initialize favoriteStateList with favorites collapsed by default
+        if (favoriteStateList.size() != favoritesMap.size()) {
+            SeptaServiceFactory.getFavoritesService().deleteAllFavoriteStates(getContext());
+            Log.d(TAG, "Reinitializing favorite states now...");
+
+            for (Map.Entry<String, Favorite> entry : favoritesMap.entrySet()) {
+                FavoriteState favoriteState = new FavoriteState(entry.getKey());
+                favoriteStateList.add(favoriteState);
+                SeptaServiceFactory.getFavoritesService().addFavoriteState(getContext(), favoriteState);
```

Comment:
I think you should be managing the Favorite states as a single list with a single method that takes all your favorite states as a list and persists them.  This is going to help you when you do drag and drop and have to maintain order.

Reply:
That is what I'm doing but I can't completely axe the favoritesMap or else people who upgrade from previous versions will lose all their favorites.

Reply:
You still have the map, but the order should be managed by a separate list of Keys.  My point was about storing the favorite state and that you should not make a service call for each state and instead save them as 1 list.